llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -2553,6 +2553,47 @@ void CombinerHelper::applyCombineTruncOfExt(
   MI.eraseFromParent();
 }
 
+bool CombinerHelper::matchCombineShlOfAnd(MachineInstr &MI, Register &Reg) {
+  // We're trying to match the following pattern:
+  //   %t = G_AND %x, (2^imm - 1)
```

Comment:
nit: I think this can be generalized to include G_SEXT, G_ZEXT, etc, anything that only touches the upper wordsize-imm bits.

Reply:
`zext(trunc x) ->x ` combine already exist. `sext(trunc x) -> x` combine is [here](38df6bc9e17ec40811acc994e4b3ce7e4bed0d98)